File: Data Structures & Algorithms/burst-balloons/submission-1.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        n = len(nums)

        memo = {}
        # dfs(left, right) means maximum coins you can get by bursting all ballons strictly between left and right
        def dfs(left, right):
            if left + 1 == right:
                return 0
            
            if (left, right) in memo:
                return memo[(left, right)]
            
            best = 0
            for i in range(left + 1, right):
                coins = nums[left] * nums[i] * nums[right]
                total = dfs(left, i) + coins + dfs(i, right)
                best = max(best, total)
            
            memo[(left, right)] = best
            return best
        
        return dfs(0, n - 1)

File: Data Structures & Algorithms/maximum-subarray/submission-1.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int curr_max = nums[0];
        int global_max = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            curr_max = max(nums[i], curr_max + nums[i]);
            global_max = max(global_max, curr_max);
        }

        return global_max;
    }
};
